### ISSUE TITLE
SimpleMLCreateModelResource when using TF-DF models

### ISSUE THREAD

**Issue by picousse** [tensorflow/serving#1865]
Hi,
Kind of a followup on:
https://github.com/tensorflow/decision-forests/issues/14

I get the same error:
```
2021-06-01 12:17:30.653126: I tensorflow_serving/core/loader_harness.cc:155] Encountered an error for servable version {name: model version: 1}: Not found: Op type not registered 'SimpleMLCreateModelResource' in binary running on 5f7b84545a3b. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed.
```

In the above the issue the solution has been defined, however it is unusable in the dockhub posted containers?

BR,
Pieter

**Comment by picousse** [comment]
HI,
I've been following https://www.tensorflow.org/tfx/serving/custom_op in an attempt to add the OP's of Tensorflow decision forests.
I copied everything from https://github.com/tensorflow/decision-forests/blob/main/tensorflow_decision_forests/tensorflow/ops/inference/ to `custom_ops/tfdf`.

Folder structure:

![image](https://user-images.githubusercontent.com/5084152/120491390-7c7f0f80-c3b9-11eb-87a6-6959e78510f5.png)


I modified the included `BUILD` file, according to the first mentioned guide.
I added : 

```
cc_library(
    name = 'tfdf',
    srcs = [
        "op.cc",
    ],
    alwayslink = 1,
    deps = [
        "@org_tensorflow//tensorflow/core:framework",
    ]
)
```
To the https://github.com/tensorflow/decision-forests/blob/main/tensorflow_decision_forests/tensorflow/ops/inference/BUILD file.
I'm not sure whether I should delete all the rest as well?

I then added ` "//tensorflow_serving/custom_ops/td_df:tfdf",` SUPPORTED_TENSORFLOW_OPS

And finally I ran `tools/run_in_docker.sh bazel build tensorflow_serving/model_servers:tensorflow_model_server`.

However this failed. What might I be doing wrong?

**Comment by picousse** [comment]
I'm now bumping into 
```
ERROR: <builtin>: Couldn't build file stable-status.txt: Failed to determine workspace status: Process exited with status 127
/bin/sh: 1: /proc/self/cwd/tools/gen_status_stamp.sh: not found
```

The build runs further, but as far as I  can see it is the only error the script encounters...
I've made the file executable.

**Comment by picousse** [comment]
Is it possible this is a shell problem?

```
/bin/sh  gen_status_stamp.sh
```
yields

```
gen_status_stamp.sh: 30: [[: not found
BUILD_SCM_REVISION b5ca0c68
```

whilst
```
/bin/bash  gen_status_stamp.sh
```
yields
```
BUILD_SCM_REVISION b5ca0c68
```

**Comment by picousse** [comment]
Got passed the error by removing:
```
  #if [[ $? != 0 ]];
  #then
  #    exit 1
  #fi
```

from gen_status_stamp.sh .

However the necessary parts are still not loaded...

```
2021-06-02 19:37:08.320471: E tensorflow_serving/util/retrier.cc:37] Loading servable: {name: penguin_model version: 1} failed: Not found: Op type not registered 'SimpleMLCreateModelResource' in binary running on builder-3. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed.
```

**Comment by picousse** [comment]
Some more work.
I didn't load in the the kernel of the OP.

I modified the

```
cc_library(
    name = 'tfdf',
    srcs = [
        "op.cc",
        "kernel.cc"
    ],
    alwayslink = 1,
    deps = [
        "@org_tensorflow//tensorflow/core:framework",
        "@com_google_absl//absl/status",
        "@com_google_absl//absl/strings",
        "@org_tensorflow//tensorflow/core:framework_headers_lib",
        "@ydf//yggdrasil_decision_forests/dataset:data_spec_cc_proto",
        "@ydf//yggdrasil_decision_forests/dataset:vertical_dataset",
        "@ydf//yggdrasil_decision_forests/model:abstract_model",
        "@ydf//yggdrasil_decision_forests/model:model_library",
        "@ydf//yggdrasil_decision_forests/utils:compatibility",
        "@ydf//yggdrasil_decision_forests/utils:distribution_cc_proto",
        "@ydf//yggdrasil_decision_forests/utils:status_macros",
        "@ydf//yggdrasil_decision_forests/utils:tensorflow",
    ]
)

```

Unfortuntaly I still ran in to the same problem:
```
2021-06-02 19:57:11.832391: E tensorflow_serving/util/retrier.cc:37] Loading servable: {name: penguin_model version: 1} failed: Not found: Op type not registered 'SimpleMLCreateModelResource' in binary running on builder-3. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed.
```

Is there a way to see all loaded OP's?

**Comment by picousse** [comment]
Found an error in the Build file.
I added the OP to the wrong TF version :-S

```serving/tensorflow_serving/model_servers/BUILD
SUPPORTED_TENSORFLOW_OPS = if_v2([ "//tensorflow_serving/custom_ops/tfdf:tfdf",  ]) + if_not_v2([
    "@org_tensorflow//tensorflow/contrib:contrib_kernels",
    "@org_tensorflow//tensorflow/contrib:contrib_ops_op_lib",
    "//tensorflow_serving/custom_ops/tfdf:tfdf",
]) + [
    "@org_tensorflow_text//tensorflow_text:ops_lib",
]
```

I've also added the decision trees to the third party section:

```
third_party/
├── absl
│   ├── BUILD
│   └── workspace.bzl
├── icu
│   ├── BUILD
│   ├── README
│   └── data.patch
├── libevent
│   └── BUILD
├── protobuf
│   ├── BUILD
│   └── workspace.bzl
├── rapidjson
│   └── BUILD
├── tf_text
│   ├── BUILD
│   └── tftext.patch
└── yggdrasil_decision_forests
    ├── BUILD
    └── workspace.bzl
```

```yggdrasil_decision_forests/BUILD

licenses(["notice"])
```

```yggdrasil_decision_forests/workspace.bzl
"""Yggdrasil Decision Forests project."""

load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

def deps():
    http_archive(
        name = "ydf",
        urls = ["https://github.com/google/yggdrasil-decision-forests/archive/refs/heads/main.zip"],
        strip_prefix = "yggdrasil-decision-forests-main",
    )

    # You can also clone the YDF repository manually.
    #native.local_repository(
    #   name = "ydf",
    #   path = "../yggdrasil_decision_forests_bazel",
    #)
```

Finally I also modified the WORKSPACE file to import these third party parts.

```WORKSPACE


# Third party libraries
#load("//third_party/absl_py:workspace.bzl", absl_py = "deps")
load("//third_party/absl:workspace.bzl", absl = "deps")
#load("//third_party/gtest:workspace.bzl", gtest = "deps")
load("//third_party/protobuf:workspace.bzl", protobuf = "deps")

absl()
#absl_py()
#gtest()
protobuf()

# Yggdrasil Decision Forests
load("//third_party/yggdrasil_decision_forests:workspace.bzl", yggdrasil_decision_forests = "deps")
yggdrasil_decision_forests()

load("@ydf//yggdrasil_decision_forests:library.bzl", ydf_load_deps = "load_dependencies")
ydf_load_deps(
    exclude_repo = [
        "absl",
        "protobuf",
        "zlib",
    ],
    repo_name = "@ydf",
)
```

When building, the OP gets picked up, as far as I can see.
Some of the ydf files however don't get compiled because of 
```
 #include <filesystem>
          ^~~~~~~~~~~~
```

Tips?

**Comment by picousse** [comment]
I switched to compiling in C++17 as that was required for the  yggdrasil library.

I got the whole thing compiled and I'm able to run the model server!

I see the model being loaded properly:
```
Model loaded with 300 root(s), 3936 node(s), and 7 input feature(s).
Successfully loaded servable version {name: penguin_model version: 1}
```
and so.
Meta data query of the model works.

However, when I want to do an inference call I directly ran into:
```
  what():  std::bad_alloc
start_direct_model_server: line 1: 121064 Aborted                 (core dumped) bazel-bin/tensorflow_serving/model_servers/tensorflow_model_server -v=4 --rest_api_port=9000 --model_name=penguin_model --model_base_path="gs://xxxxxxxxxx-ai_platform/penguin_model"
```

Verbose is on 4 but doesn't do much so it seems...
Tips?

**Comment by picousse** [comment]
@julienschuermans

**Comment by achoum** [comment]
Hi @picousse,

Following are (some of) the answers:

> I switched to compiling in C++17 as that was required for the yggdrasil library.

**TLDR**

Compile in C++14 and with the flag `--define=use_tensorflow_io=1`.

**Details**

TF does not support (yet) C++17 yet, instead it needs to be compiled with C++14.
Therefore, in TF-DF, Yggdrasil also have to be compiled with C++14.

`#include <filesystem>` is only available in C++17. Yggdrasil does not use it `use_tensorflow_io=1` ([details](https://github.com/google/yggdrasil-decision-forests/blob/main/yggdrasil_decision_forests/utils/BUILD#L54))

> Other Flags

There is a couple of others flags that might be interesting to you. See the [tf-df bazelrc](https://github.com/tensorflow/decision-forests/blob/main/.bazelrc). You probably want at least:

build --define=use_tensorflow_io=1




> std::bad_alloc

**Comment by picousse** [comment]
Thanks for the help @achoum.

As you have probably noticed, the build worked out, and I have opened a PR.

Meanwhile, I also have put a custom docker image online, that can be used for serving these TF-DF models.

#1887

https://hub.docker.com/repository/docker/ml6team/tf-serving-tfdf

https://blog.ml6.eu/serving-decision-forests-with-tensorflow-b447ea4fc81c

Thanks for the help!

Pieter

**Comment by Koushik667** [comment]
Hi @picousse  i am getting the same errror when i save the model in python using model.save and then load the model using tensorflow C API. Can you please let me know if we can use tensorflow decsion forest model in C API?

**Comment by picousse** [comment]
I'm not sure if I really understand what you mean?
So basically you save the model using the python package, and now you want to use the TensorFlow C API? 
What goal do you have in mind?

I only worked on TFX serving, helping to serve TF-DF models.

I would guess that the necessary OPS parts are missing in the Tensorflow C API?
Could you give me a dummy protocol I could follow? If I find some time, I can have a look, albeit I'm not the Tensorflow expert you are probably looking for (haven't used the C API).
But I can have a look if the fixes I did work for TensorFlow C  API as well

**Comment by Koushik667** [comment]
> I'm not sure if I really understand what you mean?
So basically you save the model using the python package, and now you want to use the TensorFlow C API?
What goal do you have in mind?

I have trained the model in python using tfdf. I want to be able to do the inference in c code, i am able to do this for boosted tree estimator in tensorflow.

> Could you give me a dummy protocol I could follow? 

i have followed the below guide to load and run the model in c.
[undocumented-tensorflow-c-api](https://medium.com/@vladislavsd/undocumented-tensorflow-c-api-b527c0b4ef6)

i have installed official tensorflow c api from [here](https://www.tensorflow.org/install/lang_c)

**Comment by shayansadeghieh** [comment]
@achoum I am running into a similar issue. There has been good work done [here](https://github.com/tensorflow/serving/pull/1887) by picousse, but he seems to be stuck on the tests. Do you mind taking a look? Thank you!

**Comment by achoum** [comment]
Hi,

> @shayansadeghieh 

I wrote [a guide](https://github.com/tensorflow/decision-forests/blob/main/documentation/tensorflow_serving.md) to help with the compilation of TF-Serving for TensorFlow Decision Forests. Do you mind checking if this works for you?

We (the TF-DF team) are currently not publishing the docker image, but maybe we will do it in the future.

> @picousse 

Thanks for your work [[1](https://github.com/tensorflow/serving/pull/1887), [2](https://blog.ml6.eu/serving-decision-forests-with-tensorflow-b447ea4fc81c), [3](https://github.com/picousse/tensorflow-serving-tfdf)].

Regarding [1](https://github.com/tensorflow/serving/pull/1887), it seems unlikely the TF-Serving team accept to include the op. The tendency in TF is to reduce as much as possible what are the core ops. 

Regarding [3](https://github.com/picousse/tensorflow-serving-tfdf), the code is relatively similar to what the guide propose. The main difference being that TF-DF is injected as a dependency instead of being copied in the TF-Serving repo.

> @Koushik667

Regarding the questions about serving with TF CC API, if possible, I would use the Yggdrasil Decision Forests API instead ([documentation](https://github.com/google/yggdrasil-decision-forests/blob/main/documentation/user_manual.md#model-inference), [example](https://github.com/google/yggdrasil-decision-forests/blob/main/examples/standalone/beginner.cc#L159)]. The API should be simpler to use, faster to run (no tensorflow overhead) and faster to compile (YDF takes ~10 minutes to compile entirely).

If you want to run the inference in C++ using TensorFlow, I would try to use the [SavedModel CC API](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/cc/saved_model/loader.h) first (since a TF-DF model is a SavedModel). While it is not more documented that the raw CC API, it is a bit easier to use :).

**Comment by picousse** [comment]
Alright, looks really good. Having TF-DF injected as a dependency is really neat, and more future proof compared to my solution.
Happy to have helped.

Hope I can help more in the future.

**Comment by UsharaniPagadala** [comment]
@picousse 

Could you please confirm If this issue can be closed based on this [comment](https://github.com/tensorflow/serving/pull/1887#issuecomment-964890525).Thanks

**Comment by UsharaniPagadala** [comment]
@picousse 
Closing this issue.Please feel free to reopen if this still persists.Thanks

**Comment by Arnold1** [comment]
@picousse @UsharaniPagadala does the c api () also support decision-forests?

libtensorflow-cpu-linux-x86_64-2.9.1.tar.gz:
```
# Install Tensorflow for C: https://www.tensorflow.org/install/lang_c
ENV FILENAME libtensorflow-cpu-linux-x86_64-2.9.1.tar.gz
RUN wget -c https://storage.googleapis.com/tensorflow/libtensorflow/${FILENAME}
RUN tar -C /usr/local -xzf ${FILENAME} 
RUN ldconfig /usr/local/lib
```

can I use the c api from the link above?

**Comment by picousse** [comment]
I can't help you. I've worked on this, mainly to get a docker container running, but ever since TF-serving took over, I haven't worked further on this.

**Comment by garciajg** [comment]
@picousse so using your image I still get this error. Have you ran into it before?

```
Loading servable: {name: forest_tree version: 3} failed: Unknown: TensorFlow: INVALID_ARGUMENT: Cannot parse binary proto from /models/forest_tree/3/assets/header.pb
         [[{{node simple_ml/SimpleMLLoadModelFromPathWithHandle}}]]
```
